### PR TITLE
Avoid depending on sizeof(T) for ExplicitLayout types due to Mono bug

### LINKED ISSUE #23722: Checksum throws on Mono 64-bit
**Version Used**: 2.6.0

The latest Roslyn release results in test failures in OmniSharp on Linux/OSX (failing build log: [omnisharp-build-log.txt](https://github.com/dotnet/roslyn/files/1548527/omnisharp-build-log.txt)) that look so:

```
[39;49m[31m    OmniSharp.Roslyn.CSharp.Tests.RenameFacts.Rename_DoesNotUpdatesWorkspace [FAIL]
[39;49m[37m      System.AggregateException : One or more errors occurred.
[39;49m[37m      ---- System.ArgumentException : checksum must be a SHA-1 hash
[39;49m[37m      Parameter name: checksum
[39;49m[30m      Stack Trace:
[39;49m[37m          at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <c9f8153c41de4f8cbafd0e32f9bf6b28>:0 
[39;49m[37m          at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <c9f8153c41de4f8cbafd0e32f9bf6b28>:0 
[39;49m[37m          at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <c9f8153c41de4f8cbafd0e32f9bf6b28>:0 
[39;49m[37m          at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <c9f8153c41de4f8cbafd0e32f9bf6b28>:0 
[39;49m[37m          at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in <c9f8153c41de4f8cbafd0e32f9bf6b28>:0 
[39;49m[37m          at Microsoft.CodeAnalysis.FindSymbols.SyntaxTreeIndex+<GetChecksumAsync>d__53.MoveNext () [0x00087] in <82fd95a6d9bb4f6e9a2da5a34e8280f2>:0 
[39;49m[37m        --- End of stack trace from previous location where exception was thrown ---

<snip lots of async state machine transitions>

[39;49m[37m        ----- Inner Stack Trace -----
[39;49m[37m          at Microsoft.CodeAnalysis.Checksum..ctor (System.Byte[] checksum) [0x00031] in <82fd95a6d9bb4f6e9a2da5a34e8280f2>:0 
[39;49m[37m          at Microsoft.CodeAnalysis.Checksum.ComputeChecksum (System.IO.Stream stream, System.Security.Cryptography.IncrementalHash hash) [0x00042] in <82fd95a6d9bb4f6e9a2da5a34e8280f2>:0 
[39;49m[37m          at Microsoft.CodeAnalysis.Checksum.Create (System.IO.Stream stream) [0x0000b] in <82fd95a6d9bb4f6e9a2da5a34e8280f2>:0 
[39;49m[37m          at Microsoft.CodeAnalysis.Checksum.Create (Microsoft.CodeAnalysis.Serialization.WellKnownSynchronizationKind kind, Roslyn.Utilities.IObjectWritable object) [0x00024] in <82fd95a6d9bb4f6e9a2da5a34e8280f2>:0 
[39;49m[37m          at Microsoft.CodeAnalysis.ProjectInfo+ProjectAttributes.Microsoft.CodeAnalysis.IChecksummedObject.get_Checksum () [0x0000e] in <82fd95a6d9bb4f6e9a2da5a34e8280f2>:0 
[39;49m[37m          at Microsoft.CodeAnalysis.Serialization.Serializer.CreateChecksum (System.Object value, System.Threading.CancellationToken cancellationToken) [0x00028] in <82fd95a6d9bb4f6e9a2da5a34e8280f2>:0 
[39;49m[37m          at Microsoft.CodeAnalysis.ProjectState+<ComputeChecksumsAsync>d__111.MoveNext () [0x000ce] in <82fd95a6d9bb4f6e9a2da5a34e8280f2>:0 
```

It appears that this is due to a struct layout padding issue on Mono 64-bit, which was found and "fixed" by https://github.com/dotnet/roslyn/pull/22794, but not in master. This is currently blocking OmniSharp, and thus, VS Code from taking a new Roslyn and adding support for C# 7.2.

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #23722

<details><summary>Ask Mode template not completed</summary>

### Customer scenario

What does the customer do to get into this situation, and why do we think this
is common enough to address for this release.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

### Bugs this fixes

(either VSO or GitHub links)

### Workarounds, if any

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

### Risk

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

### Performance impact

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

### Root cause analysis

How did we miss it?  What tests are we adding to guard against it in the future?

### How was the bug found?

(E.g. customer reported it vs. ad hoc testing)

### Test documentation updated?

If this is a new non-compiler feature or a significant improvement to an existing feature, update https://github.com/dotnet/roslyn/wiki/Manual-Testing once you know which release it is targeting.

</details>

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
We already have #22794 merged which addressed this issue for VS for Mac. Why do we need this one as well? Or should we undo the other fix?

**Comment by sharwell** [comment]
@jasonmalinowski Technically #22794 fails to meet the following condition from ECMA-335 §II.22.8 (ClassLayout):

> If *Parent* indexes an *ExplicitLayout* type, then ... *PackingSize* shall be 0. (It makes no sense to provide explicit offsets for each field, as well as a packing size.) [ERROR]

The approach in this pull request adheres to the condition while still not relying on the CLI implementation to adhere to the specification with respect to explicitlayout types.

**Comment by jasonmalinowski** [comment]
Should we still leave the other fix in? That fix at least means the memory usage is smaller. I can imagine we're OK saying they complement each other.

**Comment by sharwell** [comment]
@jasonmalinowski My intent with this is to replace the other fix. The memory usage under 64-bit would be silently corrected in the future when Mono fixes the underlying bug.

**Comment by Therzok** [comment]
cc @vargaz @kumpera

**Comment by Therzok** [comment]
Seems like a PR is up: https://github.com/mono/mono/pull/6223/files

**Comment by vargaz** [comment]
It will take some time for that PR to show up in xamarin releases.

**Comment by DustinCampbell** [comment]
I'm super-happy to see the Mono PR (thanks @vargaz!), but the workaround will still be needed in the near future. Currently, OmniSharp requires Mono 5.2.0 or higher and moving that requirement to some future Mono version would be very disruptive for VS Code customers.

**Comment by natidea** [comment]
Approved to merge via https://devdiv.visualstudio.com/DevDiv/_workitems/edit/536826

**Comment by sharwell** [comment]
@dotnet-bot retest ubuntu_16_debug_prtest please

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs [hunk 1]
```diff
@@ -123,6 +133,15 @@ public void WriteTo(ObjectWriter writer)
                 writer.WriteInt32(Data3);
             }
 
+            public static unsafe Sha1Hash FromPointer(Sha1Hash* hash)
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs]: Is there some way to use ref returns here?

**Comment by DustinCampbell** [comment]
FYI: @jaredpar and I are looking at the ubuntu_16 failure. For some reason, it's picking up master rather than dev15.5.x.

**Comment by DustinCampbell** [comment]
looks like it fixed itself.

**Comment by jaredpar** [comment]
The infrastructure is afraid of me. Once I was mentioned it got scared and decided to behave nicely.

**Pull request merged**